components/sup/src/ctl_gateway/codec.rs

```diff
@@ -0,0 +1,504 @@
+//!                        message.
+//!     * flags (5-bit) - reserved
+//!     * message_id_len (6-bit) - length of the message ID segment in bytes
+//!     * body_len (20-bits) - length of the message body segment in bytes
+//!
+//! # Transaction Segment
+//!
+//! An optional message segment containing transaction information for requests and responses.
+//! This segment is encoded or decoded only if the `is_txn` bit is set on the header segment. The
+//! transaction segment is a 32 bit integer in big-endian format and is packed as follows:
+//!
+//! * Transaction Segment (32-bit)
+//!     * is_response (1-bit) - flag to determine if the message is a response to a request. If
+//!                             not set it is a request, if set it is a reply.
+//!     * is_complete (1-bit) - flag to determine if the message is the last in a stream.
+//!     * txn_identifier (30-bit) - the actual transaction identifier. A number between 0 and
+//!                                 `TXN_ID_MASK` (2^30-1).
+//!
+//! # Message ID Segment
+//!
+//! Used as a hint to the decoder for which protocol message is contained in the following message
+//! body segment. The message ID segment is variable length indicated by the `message_id_len` field
+//! on the header segment. It contains a string representation of the message ID.
+//!
+//! # Message Body Segment
+//!
+//! Contains the actual payload of the message encoded using Google
+//! [Protobuf 2](https://developers.google.com/protocol-buffers/docs/reference/proto2-spec).
+
+use std::fmt;
+use std::io::{self, Cursor};
+
+use bytes::{BigEndian, Buf, BufMut, Bytes, BytesMut};
+use futures;
+use protobuf::{self, MessageStatic};
+use tokio::net::TcpStream;
+use tokio_io::codec::{Decoder, Encoder, Framed};
+
+use protocols::net::{NetErr, NetResult};
+
+const BODY_LEN_MASK: u32 = 0xFFFFF;
+const HEADER_LEN: usize = 4;
+const MESSAGE_ID_MASK: u32 = 0x3F;
+const MESSAGE_ID_OFFSET: u32 = 20;
+const TXN_LEN: usize = 4;
+const TXN_OFFSET: u32 = 31;
+
+const TXN_ID_MASK: u32 = 0x3FFFFFFF;
+const RESPONSE_OFFSET: u32 = 31;
+const RESPONSE_MASK: u32 = 0x1;
+const COMPLETE_OFFSET: u32 = 30;
+const COMPLETE_MASK: u32 = 0x1;
+
+/// A `TcpStream` framed with `SrvCodec`. This is the base socket connection that the CtlGateway
+/// client and server speak.
+pub type SrvStream = Framed<TcpStream, SrvCodec>;
+
+/// Sending half of `SrvStream`.
+pub type SrvSink = futures::stream::SplitSink<SrvStream>;
+
+/// An unsigned 32-bit integer packed with transaction information which is present if a request
+/// should receive a response from the destination.
+#[derive(Default, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct SrvTxn(u32);
+
+impl SrvTxn {
+    /// The contained transaction ID.
+    pub fn id(&self) -> u32 {
+        self.0 & TXN_ID_MASK
+    }
+
+    /// Generates and returns the next valid transaction ID for the given transaction ID.
```

Comment:
This comment needs updating since this now mutates self rather than returning anything

Reply:
Got it :)